File: frontend/src/stores/conversations.ts
import { create } from "zustand";
import type { ConversationPayload } from "@/api/tauri";
import {
  getConversations,
  createGroupConversation,
  addToConversation,
} from "@/api/tauri";

interface ConversationsState {
  conversations: ConversationPayload[];
  activeConversationId: string | null;
  loading: boolean;

  loadConversations: () => Promise<void>;
  selectConversation: (id: string) => void;
  createGroup: (peerIds: string[], name?: string) => Promise<void>;
  addParticipant: (convId: string, peerId: string) => Promise<void>;
}

export const useConversationsStore = create<ConversationsState>((set) => ({
  conversations: [],
  activeConversationId: null,
  loading: false,

  loadConversations: async () => {
    set({ loading: true });
    try {
      const conversations = await getConversations();
      set({ conversations, loading: false });
    } catch (err) {
      console.warn("Failed to load conversations:", err);
      set({ loading: false });
    }
  },

  selectConversation: (id) => {
    set({ activeConversationId: id });
  },

  createGroup: async (peerIds, name) => {
    try {
      const conv = await createGroupConversation(peerIds, name);
      set((state) => ({
        conversations: [...state.conversations, conv],
        activeConversationId: conv.id,
      }));
    } catch (err) {
      console.error("Failed to create group conversation:", err);
    }
  },

  addParticipant: async (convId, peerId) => {
    try {
      await addToConversation(convId, peerId);
      set((state) => ({
        conversations: state.conversations.map((c) =>
          c.id === convId
            ? { ...c, participants: [...c.participants, peerId] }
            : c,
        ),
      }));
    } catch (err) {
      console.error("Failed to add participant:", err);
    }
  },
}));

File: src-tauri/src/commands/conversations.rs
use chrono::Utc;
use serde::Serialize;
use uuid::Uuid;

use concord_core::types::DirectConversation;

use crate::AppState;

/// JSON-serializable conversation payload for the frontend.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConversationPayload {
    pub id: String,
    pub participants: Vec<String>,
    pub created_at: i64,
    pub is_group: bool,
    pub name: Option<String>,
}

impl From<&DirectConversation> for ConversationPayload {
    fn from(conv: &DirectConversation) -> Self {
        Self {
            id: conv.id.clone(),
            participants: conv.participants.clone(),
            created_at: conv.created_at.timestamp_millis(),
            is_group: conv.is_group,
            name: conv.name.clone(),
        }
    }
}

/// Get all conversations.
#[tauri::command]
pub fn get_conversations(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<ConversationPayload>, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let conversations = db.get_conversations().map_err(|e| e.to_string())?;
    Ok(conversations.iter().map(ConversationPayload::from).collect())
}

/// Create a group conversation.
#[tauri::command]
pub fn create_group_conversation(
    state: tauri::State<'_, AppState>,
    peer_ids: Vec<String>,
    name: Option<String>,
) -> Result<ConversationPayload, String> {
    let mut participants = peer_ids;
    if !participants.contains(&state.peer_id) {
        participants.push(state.peer_id.clone());
    }

    let conv = DirectConversation {
        id: Uuid::new_v4().to_string(),
        participants,
        created_at: Utc::now(),
        is_group: true,
        name,
    };

    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.create_conversation(&conv).map_err(|e| e.to_string())?;

    Ok(ConversationPayload::from(&conv))
}

/// Add a peer to an existing conversation.
#[tauri::command]
pub fn add_to_conversation(
    state: tauri::State<'_, AppState>,
    conversation_id: String,
    peer_id: String,
) -> Result<(), String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.add_participant(&conversation_id, &peer_id)
        .map_err(|e| e.to_string())
}
